### ISSUE TITLE
Race condition after/during compile

### ISSUE THREAD

**Issue by oskarkook** [phoenixframework/phoenix#3494]
Reproducible with the following steps:
- Create new project with default configuration: `mix phx.new phoenix_test`
- Run server with `mix phx.server`
- Make a change to `PageController` file, no live reload should be triggered
- When you manually refresh page, a compile will be triggered and initial Live Reload connection to websockets will fail:
```
Compiling 2 files (.ex)
[error] #PID<0.771.0> running PhoenixTestWeb.Endpoint (connection #PID<0.770.0>, stream id 1) terminated
Server: localhost:4000 (http)
Request: GET /phoenix/live_reload/socket/websocket?vsn=2.0.0
** (exit) an exception was raised:
    ** (UndefinedFunctionError) function PhoenixTestWeb.Endpoint.__handler__/2 is undefined (module PhoenixTestWeb.Endpoint is not available)
        (phoenix_test) PhoenixTestWeb.Endpoint.__handler__(%Plug.Conn{adapter: {Plug.Cowboy.Conn, :...}, assigns: %{}, before_send: [], body_params: %Plug.Conn.Unfetched{aspect: :body_params}, cookies: %Plug.Conn.Unfetched{aspect: :cookies}, halted: false, host: "localhost", method: "GET", owner: #PID<0.771.0>, params: %Plug.Conn.Unfetched{aspect: :params}, path_info: ["phoenix", "live_reload", "socket", "websocket"], path_params: %{}, port: 4000, private: %{}, query_params: %Plug.Conn.Unfetched{aspect: :query_params}, query_string: "vsn=2.0.0", remote_ip: {127, 0, 0, 1}, req_cookies: %Plug.Conn.Unfetched{aspect: :cookies}, req_headers: [{"accept", "*/*"}, {"accept-encoding", "gzip, deflate"}, {"accept-language", "en-US,en;q=0.5"}, {"cache-control", "no-cache"}, {"connection", "keep-alive, Upgrade"}, {"cookie", "_constructor_key=SFMyNTY.g3QAAAADbQAAAAtfY3NyZl90b2tlbm0AAAAYSFJrZUpqNGtoWUhqcGhtejRRSmxmcWkybQAAAAZsb2NhbGVtAAAABWV0X0VFbQAAAAd1c2VyX2lkbQAAAAEx.PmfgcXUbunsM0lzoDZWvPgG1jE6M4h-nOljf7afTUUk; locale=et_EE; _demo_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYVklvaHAydlc4cm5DbDh0YnZqbU5qRWl4.ku1noT-GCAR1rHe-ty_51ApsqrQ500gLEP7JnKdMHKE"}, {"host", "localhost:4000"}, {"origin", "http://localhost:4000"}, {"pragma", "no-cache"}, {"sec-websocket-extensions", "permessage-deflate"}, {"sec-websocket-key", "7ZGb2eUpSoM03ZP4+r8dJA=="}, {"sec-websocket-version", "13"}, {"upgrade", "websocket"}, {"user-agent", "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0"}], request_path: "/phoenix/live_reload/socket/websocket", resp_body: nil, resp_cookies: %{}, resp_headers: [{"cache-control", "max-age=0, private, must-revalidate"}], scheme: :http, script_name: [], secret_key_base: nil, state: :unset, status: nil}, [])
        (phoenix) lib/phoenix/endpoint/cowboy2_handler.ex:17: Phoenix.Endpoint.Cowboy2Handler.init/2
        (cowboy) /home/oskar/Development/phoenix_test/deps/cowboy/src/cowboy_handler.erl:41: :cowboy_handler.execute/2
        (cowboy) /home/oskar/Development/phoenix_test/deps/cowboy/src/cowboy_stream_h.erl:296: :cowboy_stream_h.execute/3
        (cowboy) /home/oskar/Development/phoenix_test/deps/cowboy/src/cowboy_stream_h.erl:274: :cowboy_stream_h.request_process/3
        (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
[info] GET /
[debug] Processing with PhoenixTestWeb.PageController.index/2
  Parameters: %{}
  Pipelines: [:browser]
[info] Sent 200 in 263µs
```

WS connection will be established after first failure, so Live Reload still works, but these exceptions make it really annoying to parse the log. With Live View the issue is worse because you can end up with a blank page altogether.

`mix deps`: https://gist.github.com/oskarkook/b70ac5fbb995e2fb353f7934a5ae67de
Initial report here: https://github.com/phoenixframework/phoenix_live_reload/issues/97

### Environment

* Elixir version (elixir -v): 1.9.0
* Phoenix version (mix deps): 1.4.9
* NodeJS version (node -v): 10.16.0
* NPM version (npm -v): 6.9.0
* Operating system: Fedora 30

**Comment by oskarkook** [comment]
@josevalim

**Comment by DavidVII** [comment]
[I have the same issue!](https://elixirforum.com/t/rosweb-endpoint-module-not-available-after-compiling/22181) I can't reproduce it on a fresh install, but I get this error **constantly** on a project I've been working on for a few months now. It happens whenever I edit a template file and the reloader kicks in. The websocket connections return as not found in the browser's console too whenever the page reloads.

I originally thought this might have to do with Phoenix LiveView because I started noticing this issue around the same time I added it to my project. This definitely comes across as a race condition, but I have no idea how to go about hunting this bug down.

I'm happy to jump on a screen share and share any details about my project to help narrow it down with anyone who might have an idea on a possible fix.
